File: autozyme_py/src/autozyme/_verify_worker.py
"""Subprocess worker for verify_patch — one measurement per invocation.

Spawned by `autozyme._verify._verify_one_tier` once for baseline and once for
patched, per rep. The parent runs:

    python -m autozyme._verify_worker
        --patch <name>
        --task-dir <abs/path>
        --tier <name>
        --output-dir <abs/path>
        [--activate]

Inside the subprocess we:
  1. Import the patch submodule (registers it; needed even for baseline since
     baseline still calls `patch.smoke[...]`).
  2. If --activate: monkey-patch upstream via `autozyme.activate(name)`.
  3. Build inputs via `patch.smoke["load"](task_dir, tier)`        — untimed.
  4. Time only `result = patch.smoke["call"](inputs)`              — the timed
     region. Use time.perf_counter().
  5. Save outputs via `patch.smoke["save"](result, output_dir, tier=tier)`.
  6. Print a single JSON line on stdout: {"elapsed_sec": <float>}.

Rationale: in-process deactivate() → baseline_call() → activate() → patched_call()
required deepcopy + Pyro param-store reset / TF seed-reset code inside
_smoke_call, which were timed and diluted the reported speedup (cell2location
5x → ~2.1x). Spawning a fresh process per measurement removes the contamination
entirely; each subprocess starts from a clean Python interpreter so smoke
recipes don't need any reset boilerplate.
"""
from __future__ import annotations

import argparse
import json
import sys
import time


def _peak_rss_mb() -> float | None:
    """Peak RSS of this subprocess, in MiB. Returns None if unavailable.

    Linux ``ru_maxrss`` is KiB, macOS is bytes. Windows lacks the ``resource``
    module — falls back to psutil if importable, else None.
    """
    try:
        import resource  # POSIX only
        maxrss = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
        if sys.platform == "darwin":
            return maxrss / (1024 * 1024)
        return maxrss / 1024
    except Exception:
        try:
            import psutil
            return psutil.Process().memory_info().peak_wset / (1024 * 1024)
        except Exception:
            return None


def _parse_args(argv: list[str]) -> argparse.Namespace:
    p = argparse.ArgumentParser(prog="python -m autozyme._verify_worker")
    p.add_argument("--patch", required=True,
                   help="registered patch name (e.g. 'cell2location')")
    p.add_argument("--task-dir", required=True,
                   help="absolute path to the autozyme task directory")
    p.add_argument("--tier", required=True,
                   help="tier name (e.g. 'tiny', 'medium', 'large')")
    p.add_argument("--output-dir", required=True,
                   help="absolute path where smoke.save() writes its outputs")
    p.add_argument("--activate", action="store_true",
                   help="activate the patch before timing (patched run)")
    return p.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = _parse_args(argv if argv is not None else sys.argv[1:])

    import autozyme
    from autozyme._core import _REGISTRY, _import_submodule

    # Always import the submodule — even baseline needs the patch's smoke
    # recipe (smoke.load/call/save live on the registered patch object).
    if args.patch not in _REGISTRY:
        _import_submodule(args.patch)
    patch = _REGISTRY.get(args.patch)
    if patch is None:
        raise RuntimeError(
            f"patch {args.patch!r} did not register on import"
        )
    if patch.smoke is None:
        raise RuntimeError(
            f"patch {args.patch!r} has no smoke recipe"
        )

    from autozyme._smoke import resolve_smoke
    smoke = resolve_smoke(args.task_dir, patch)
    if smoke is None:
        raise RuntimeError(
            f"no smoke recipe for patch {args.patch!r} "
            f"(add attest/smoke.py under task dir)"
        )

    # `zyme package check-intercept` opts in via ZYME_INSTRUMENT_INTERCEPTS=1.
    # The probe wraps each registered fast fn with a counter and writes JSON
    # to ZYME_INTERCEPT_OUT at exit. Must run AFTER the patch is registered
    # (so targets are present) and BEFORE activate() resolves dispatchers.
    from autozyme._intercept_probe import install_from_env
    install_from_env()

    if args.activate:
        autozyme.activate(args.patch)

    inputs = smoke["load"](args.task_dir, args.tier)
    t0 = time.perf_counter()
    result = smoke["call"](inputs)
    elapsed = time.perf_counter() - t0
    smoke["save"](result, args.output_dir, tier=args.tier)

    # Peak RSS over this subprocess's lifetime (load + call + save). Reported
    # alongside elapsed so the parent can store baseline/patched memory next
    # to baseline/patched timing in package_verify.tsv. NOTE: ru_maxrss is OS
    # RSS — it includes shared/mmap pages, so absolute values can disagree
    # with tracemalloc-based `peak_mb` in results.tsv. Use for *delta* between
    # baseline and patched on the same workload.
    payload = {"elapsed_sec": elapsed, "peak_mb": _peak_rss_mb()}
    # Single JSON line on stdout. Anything else (warnings, autozyme banners)
    # goes to stderr by convention. Parent parses the LAST non-empty line as
    # JSON to be robust to a stray third-party stdout print.
    print(json.dumps(payload), flush=True)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: autozyme_r/inst/verify_worker.R
# Subprocess worker for verify_patch — one measurement per invocation.
#
# Spawned by autozyme:::.run_worker once for baseline and once for patched,
# per rep. The parent runs:
#
#   Rscript <pkg>/inst/verify_worker.R \
#     --patch <name> --task-dir <abs> --tier <name> \
#     --output-dir <abs> [--activate]
#
# Inside the subprocess we:
#   1. library(autozyme); .ensure_registered(name) — needed even for
#      baseline since baseline still calls patch$smoke[[...]].
#   2. If --activate: autozyme::activate(name).
#   3. Build inputs via patch$smoke$load(task_dir, tier)        — untimed.
#   4. Time only result <- patch$smoke$call(inputs)              — the timed
#      region (Sys.time() difftime in seconds).
#   5. Save outputs via patch$smoke$save(result, output_dir, tier = tier).
#   6. Print a single JSON line on stdout: {"elapsed_sec": <float>}.
#
# Rationale: in-process restore() → baseline call → activate() → patched
# call previously required state-reset code (model reconstructions, seed
# resets) inside _smoke_call, which contaminated the timed region. Fresh
# subprocesses eliminate the contamination — smoke recipes need no reset
# boilerplate.

suppressPackageStartupMessages({
  library(autozyme)
})

autozyme:::.sync_worker_thread_options()

.parse_args <- function(argv) {
  # Rscript -e '...' --args foo ... leaks the literal --args sentinel into
  # commandArgs(trailingOnly = TRUE) on R 4.5.0 (Mac/Linux). The file-mode
  # `Rscript file.R --args foo ...` form does NOT — the script-name boundary
  # consumes it. Filter it out here so the CLI's -e spawn pattern is robust.
  argv <- argv[argv != "--args"]
  out <- list(activate = FALSE)
  i <- 1L
  while (i <= length(argv)) {
    key <- argv[i]
    if (key == "--activate") {
      out$activate <- TRUE
      i <- i + 1L
      next
    }
    if (!startsWith(key, "--")) {
      stop(sprintf("verify_worker: unexpected positional arg '%s'", key))
    }
    if (i + 1L > length(argv)) {
      stop(sprintf("verify_worker: flag '%s' missing value", key))
    }
    val <- argv[i + 1L]
    nm <- gsub("-", "_", sub("^--", "", key), fixed = TRUE)
    out[[nm]] <- val
    i <- i + 2L
  }
  required <- c("patch", "task_dir", "tier", "output_dir")
  missing_keys <- setdiff(required, names(out))
  if (length(missing_keys)) {
    stop(sprintf(
      "verify_worker: missing required flags: %s",
      paste(paste0("--", gsub("_", "-", missing_keys, fixed = TRUE)),
            collapse = ", ")
    ))
  }
  out
}

argv <- commandArgs(trailingOnly = TRUE)
args <- .parse_args(argv)

# Force-register the patch (the registry is keyed off inst/patches/<name>.R
# which is sourced lazily). Baseline still needs smoke$load/call/save.
ok <- autozyme:::.ensure_registered(args$patch)
if (!isTRUE(ok)) {
  stop(sprintf(
    "verify_worker: patch '%s' did not register (upstream missing?)",
    args$patch
  ))
}
patch <- autozyme:::.zyme_registry[[args$patch]]
task_dir <- normalizePath(args$task_dir, mustWork = TRUE)
smoke <- autozyme:::.resolve_smoke(task_dir, patch)
if (is.null(smoke)) {
  stop(sprintf(
    "verify_worker: no smoke recipe for patch '%s' (add attest/smoke.R under task dir)",
    args$patch))
}

# `zyme package check-intercept` opts in via ZYME_INSTRUMENT_INTERCEPTS=1.
# The probe wraps each registered fast fn with a counter and writes JSON to
# ZYME_INTERCEPT_OUT at session exit. Sourced AFTER the patch is registered
# (so targets exist) and BEFORE activate() resolves dispatchers.
if (Sys.getenv("ZYME_INSTRUMENT_INTERCEPTS", "") == "1") {
  install_fn <- tryCatch(
    get(".zyme_intercept_install_from_env", envir = asNamespace("autozyme")),
    error = function(e) NULL
  )
  if (!is.null(install_fn)) install_fn()
}

if (isTRUE(args$activate)) {
  autozyme::activate(args$patch)
}

# Pass `tier` to smoke$call only if its signature accepts it (back-compat
# with smoke recipes that take just (inputs)).
#
# Do not use do.call() here. do.call() embeds argument values into a literal
# call object; Seurat-family code may inspect match.call()/sys.call() downstack
# and then traverse a call containing the full materialized Seurat object. That
# turns small DoubletFinder/Seurat smoke calls into multi-minute stalls.
.call_smoke <- function(input) {
  fmls <- names(formals(smoke$call))
  if ("tier" %in% fmls) {
    smoke$call(input, tier = args$tier)
  } else {
    smoke$call(input)
  }
}

inputs <- smoke$load(task_dir, args$tier)
t0 <- Sys.time()
result <- .call_smoke(inputs)
elapsed <- as.numeric(difftime(Sys.time(), t0, units = "secs"))
smoke$save(result, args$output_dir, tier = args$tier)

# Single JSON line on stdout. Hand-roll to avoid a hard jsonlite dep — the
# payload is just one numeric field. Parent parses the last non-empty stdout
# line that looks like {"elapsed_sec": ...}.
cat(sprintf('{"elapsed_sec": %.9f}\n', elapsed))
flush.console()

# Some Seurat/uwot/sctransform sessions can hang during implicit R shutdown
# on Windows after the output has already been written. Kill this worker after
# flushing JSON; the parent treats a non-zero exit with valid JSON as success.
if (.Platform$OS.type == "windows") {
  system2("taskkill", c("/PID", as.character(Sys.getpid()), "/F"),
          stdout = FALSE, stderr = FALSE, wait = FALSE)
  Sys.sleep(60)
}

quit(save = "no", status = 0L, runLast = FALSE)
